core/src/main/java/jenkins/telemetry/impl/java11/MissingClassEvents.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2019 CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.telemetry.impl.java11;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import javax.annotation.Nonnull;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class MissingClassEvents {
+
+    /**
+     * Only 100 exceptions a day (period of telemetry)
+     */
+    @VisibleForTesting
+    /* package */ static /* final */ int MAX_EVENTS_PER_SEND = 100;
+
+    /**
+     * List of events, one per stack trace.
+     */
+    private ConcurrentHashMap<List<StackTraceElement>, MissingClassEvent> events = new ConcurrentHashMap<>(MAX_EVENTS_PER_SEND);
```

Comment:
I've tested with a pipeline running a custom step defined in a plugin throwing a CNFE using `sun.misc.BASE64Encoder`. The stack trace is the same and the event gathered has two occurrences:
**Exception in build console:**
`java.lang.ClassNotFoundException: sun.misc.BASE64Encoder
	at jenkins.util.AntClassLoader.findClassInComponents(AntClassLoader.java:1383)
	at jenkins.util.AntClassLoader.findClass(AntClassLoader.java:1336)
	at jenkins.util.AntClassLoader.loadClass(AntClassLoader.java:1083)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
	at com.cloudbees.rleon.LoadClassUsingClassCLStep.perform(LoadClassUsingClassCLStep.java:43)
Caused: java.io.IOException: Class not found: sun.misc.BASE64Encoder
	at com.cloudbees.rleon.LoadClassUsingClassCLStep.perform(LoadClassUsingClassCLStep.java:45)
	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:80)
	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:67)
	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)`

**Event gathered (occurrences=2):**
`MissingClassEvent{time='2019-05-08T17:22Z', occurrences=2, stackTrace='java.lang.ClassNotFoundException: sun.misc.BASE64Encoder
	at jenkins.util.AntClassLoader.findClassInComponents(AntClassLoader.java:1383)
	at jenkins.util.AntClassLoader.findClass(AntClassLoader.java:1336)
	at jenkins.util.AntClassLoader.loadClass(AntClassLoader.java:1083)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
	at com.cloudbees.rleon.LoadClassUsingClassCLStep.perform(LoadClassUsingClassCLStep.java:43)
	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:80)
	at org.jenkinsci.plugins.workflow.steps.CoreStep$Execution.run(CoreStep.java:67)
	at org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)`

How can I reproduce that scenario?

Reply:
If the concern is about the number of events reported and the lack of disk space on the telemetry endpoint, we could use the missing class as the "key of the set", instead of the "stack trace". But in this case, we are losing all the places where the missing class is thrown. Places where we should look at to analyze if the plugin requires some change to be java11 friendly. Therefore doing the telemetry data sent less helpful IMO.

Reply:
@MRamonLeon I might be missing something in the comment above. The stacktraces are different AFAICT.

The most simple way to invoke an issue is to put the bogus code into a library and to invoke it from different places of Jenkinsfile. IIRC in this case there will be a stacktrace entry pointing to a line number in the Pipeline script

Reply:
Are you referring to this line in the stack trace?
`at WorkflowScript.run(WorkflowScript:5)`

Then we would have to remove the number or even the line, but cloning the `Throwabe`. In such a way we avoid changing the object for later processing in the existing code. We can see how it works in a weekly and adjust it if it becomes massive.

Reply:
You shouldn't use mutable keys in a map. Might be a good idea to ensure they're all wrapped in unmodifiableList().

Reply:
The are @jvz. I use an _unmodifiableList_. See: https://github.com/jenkinsci/jenkins/pull/4003/files#diff-6e4aa59459f375657bfb070227cbdc2dR62

Reply:
Ok, not a problem then!